### ISSUE TITLE
Dashboard not working with RBAC

### ISSUE THREAD

**Issue by phcollignon** [ubuntu/microk8s#637]
Using Dashboard with default serviceaccount when RBAC is enabled has insufficient privileges ..

i.e. we get this kind of message : pods is forbidden: User "system:serviceaccount:default:default" cannot list resource "pods" in API group "" at the cluster scope

**Comment by phcollignon** [comment]
oops .. forgot to read the doc and create a dedicated user !
https://github.com/ubuntu/microk8s/blob/master/microk8s-resources/actions/enable.dashboard.sh

**Comment by ktsakalozos** [comment]
We may need to improve the experience here. Any suggestions? What would you expect to happen @phcollignon ?

**Comment by phcollignon** [comment]
Well .. I should have read the doc ! 

But I tried to use the default "dashboard-token" but it was not working neither .. maybe that one should be changed or add another one with more privileges as many users use microk8s for dev ..

Using the token from a created account as this works : 
define user and clusterole in a `dashboard-admin-serviceaccount.yaml` file : 
```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: dashboard-admin
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: dashboard-admin
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
  - kind: ServiceAccount
    name: dashboard-admin
    namespace: kube-system
```
and run ` kubectl -f apply dashboard-admin-serviceaccount.yaml`

Get the token from the dashboard admin user with : 
` kubectl get secret -n kube-system | grep dashboard-admin `
` kubectl describe secret dashboard-admin-token-XXXX -n kube-system `

Use that token in the browser login screen.
 (You can get the dashboard by the ClusterIP on default https port : 
` kubectl get service --all-namespaces | grep dashboard `
http://<DashboardClusterIP>:443 )

**Comment by phcollignon** [comment]
But why using the default serviceaccount do not have enough privileges by default ? 
I try running helm with rbac and also has issues .. system serviceaccount cannot create namespace ..
(so I also created a tiller service account with cluster-admin roles)

**Comment by phcollignon** [comment]
I created an unsecured dashboard addon in my fork.  If you find it useful, there is a pull request : https://github.com/ubuntu/microk8s/pull/639

**Comment by phcollignon** [comment]
I put my config files to access dashboard with RBAC here : https://github.com/phcollignon/microk8s-dashboard

**Comment by 9swampy** [comment]
Sheesh. Definitely would be helpful to @ktsakalozos improve the experience here. 

Merge @phcollignon #639 at the very least.

Maybe the microk8s.enable rbac could include these steps, dump out the token at the end of the enablement and advise the user to use the alternate token... or at the very least point the enabler to these steps/documentation!

**Comment by redsk** [comment]
@ktsakalozos I solved this problem by following the [doc](https://github.com/kubernetes/dashboard/blob/master/docs/user/access-control/creating-sample-user.md) and modifying it to suit `microk8s`. 

I defined the file `dashboard-admin.yaml`:
```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: admin-user
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: admin-user
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
- kind: ServiceAccount
  name: admin-user
  namespace: kube-system
```

I then applied it like so:
```
kubectl delete clusterrolebinding kubernetes-dashboard
kubectl apply -f dashboard-admin.yaml
```

Then get the token by:
```
token=$(kubectl -n kube-system get secret | grep admin-user | awk '{print $1}')
kubectl -n kube-system describe secret $token |grep token: |awk  '{print $2}'
```

It would definitely help to have these few line in the [microk8s doc](https://microk8s.io/docs/addon-dashboard)

**Comment by jbulow** [comment]
I run into this problem with MicroK8s v1.24.3 revision 3584. The solutions and docs above does not solve the problem. 

Theoutput of  `microk8s.kubectl -n kube-system get secret` does not contain any `admin-user`:

```
NAME                              TYPE                                  DATA   AGE
kubernetes-dashboard-certs        Opaque                                0      31d
microk8s-dashboard-token          kubernetes.io/service-account-token   3      31d
kubernetes-dashboard-csrf         Opaque                                1      31d
kubernetes-dashboard-key-holder   Opaque                                2      31d
```

**Comment by killermonk** [comment]
Since the docs link listed above is out of date, I'm going to add comments to help anyone else that landed here recently.

When enabling dashboard, the following text is output. I installed dashboard a month ago, so forgot about this, but saw it after I enabled rbac when I disabled/enabled dashboard again to see if it helped:

```
If RBAC is not enabled access the dashboard using the token retrieved with:

microk8s kubectl describe secret -n kube-system microk8s-dashboard-token

Use this token in the https login UI of the kubernetes-dashboard service.

In an RBAC enabled setup (microk8s enable RBAC) you need to create a user with restricted
permissions as shown in:
https://github.com/kubernetes/dashboard/blob/master/docs/user/access-control/creating-sample-user.md
```

I followed the steps in creating-sample-user.md but changed the namespace to be kube-system from kubernetes-dashboard and got everything working.

Basically the only extra step need that are not mentioned above is actually creating the secret/Bearer Token for the user.

Just in case the link changes again, you can get a short-lived Bearer Token to use on the dashboard using the following command:
```
kubectl -n kubernetes-dashboard create token admin-user
```
or a long lived token with
```
apiVersion: v1
kind: Secret
metadata:
  name: admin-user
  namespace: kube-system
  annotations:
    kubernetes.io/service-account.name: "admin-user"   
type: kubernetes.io/service-account-token
```
And retrieve it with
```
kubectl get secret admin-user -n kube-system -o jsonpath={".data.token"} | base64 -d
```
